### PR TITLE
Add Neon implementation of SIMD classes.

### PR DESCRIPTION
**Pull request by gershonho**

### CODE REVIEW

File: dlib/simd/simd4f.h [hunk 1]
```diff
@@ -83,6 +83,87 @@ namespace dlib
...
+    class simd4f
+    {
+    public:
+        typedef float type;
+
+        inline simd4f() {}
+        inline simd4f(float f) { x = vdupq_n_f32(f); }
+        inline simd4f(float r0, float r1, float r2, float r3)
+        {
+          float __attribute__((aligned (16))) data[4] = { r0, r1, r2, r3 };
```
Comment (StatusReport) [inline dlib/simd/simd4f.h]: Inconsistent indentation.

File: dlib/simd/simd4i.h [hunk 2]
```diff
@@ -384,6 +465,9 @@ namespace dlib
         int32 temp[4];
         item.store(temp);
         return temp[0]+temp[1]+temp[2]+temp[3];
+//#elif defined(DLIB_HAVE_NEON)
```
Comment (StatusReport) [inline dlib/simd/simd4i.h]: Remove (why BTW?).

File: dlib/simd/simd4i.h [hunk 3]
```diff
@@ -357,6 +436,8 @@ namespace dlib
     { 
 #ifdef DLIB_HAVE_SSE41
         return _mm_max_epi32(lhs, rhs); 
+#elif defined(DLIB_HAVE_NEON)
+		return vmaxq_s32(lhs, rhs);
```
Comment (StatusReport) [inline dlib/simd/simd4i.h]: Indentation.

File: dlib/simd/simd4i.h [hunk 4]
```diff
@@ -250,6 +311,14 @@ namespace dlib
     { 
 #ifdef DLIB_HAVE_SSE2
         return _mm_sra_epi32(lhs,_mm_cvtsi32_si128(rhs));
+#elif defined(DLIB_HAVE_NEON)
+        int32 _lhs[4]; lhs.store(_lhs);
+        return simd4i(_lhs[0]>>rhs,
+                      _lhs[1]>>rhs,
+                      _lhs[2]>>rhs,
+                      _lhs[3]>>rhs);
+		//int32x4_t result = vdupq_n_s32(0);
```
Comment (StatusReport) [inline dlib/simd/simd4i.h]: Remove (why?).

File: dlib/simd/simd4i.h [hunk 5]
```diff
@@ -234,6 +287,14 @@ namespace dlib
     { 
 #ifdef DLIB_HAVE_SSE2
         return _mm_sll_epi32(lhs,_mm_cvtsi32_si128(rhs));
+#elif defined(DLIB_HAVE_NEON)
+        int32 _lhs[4]; lhs.store(_lhs);
+        return simd4i(_lhs[0]<<rhs,
+                      _lhs[1]<<rhs,
+                      _lhs[2]<<rhs,
+                      _lhs[3]<<rhs);
+		//int32x4_t result = vdupq_n_s32(0);
```
Comment (StatusReport) [inline dlib/simd/simd4i.h]: Remove (why is this commented out?).

File: dlib/simd/simd4i.h [hunk 6]
```diff
@@ -148,7 +191,9 @@ namespace dlib
     inline simd4i operator* (const simd4i& lhs, const simd4i& rhs) 
     { 
 #ifdef DLIB_HAVE_SSE41
-        return _mm_mullo_epi32(lhs, rhs); 
+        return _mm_mullo_epi32(lhs, rhs);
```
Comment (StatusReport) [inline dlib/simd/simd4i.h:194]: This line has not changed.

File: dlib/simd/simd4i.h [hunk 7]
```diff
@@ -171,7 +216,9 @@ namespace dlib
     inline simd4i operator& (const simd4i& lhs, const simd4i& rhs) 
     { 
 #ifdef DLIB_HAVE_SSE2
-        return _mm_and_si128(lhs, rhs);
```
Comment (StatusReport) [inline dlib/simd/simd4i.h:174]: Same.

File: dlib/simd/simd4i.h [hunk 8]
```diff
@@ -117,6 +156,8 @@ namespace dlib
     { 
 #ifdef DLIB_HAVE_SSE2
         return _mm_add_epi32(lhs, rhs); 
+#elif defined(DLIB_HAVE_NEON)
+		return vaddq_s32(lhs, rhs);
```
Comment (StatusReport) [inline dlib/simd/simd4i.h]: Indentation. Please keep consistent to the file's indentation.

File: dlib/simd/simd4i.h [hunk 9]
```diff
@@ -44,6 +44,45 @@ namespace dlib
     private:
         __m128i x;
     };
+#elif defined(DLIB_HAVE_NEON)
+	class simd4i
```
Comment (StatusReport) [inline dlib/simd/simd4i.h]: Indentation.

File: dlib/simd/simd4f.h [hunk 10]
```diff
@@ -488,4 +618,3 @@ namespace dlib
 }
 
 #endif // DLIB_sIMD4F_Hh_
-
```
Comment (StatusReport) [inline dlib/simd/simd4f.h:491]: Let's do minimal changes, so remove.

File: dlib/simd/simd4f.h [hunk 11]
```diff
@@ -424,7 +546,8 @@ namespace dlib
     {
 #ifdef DLIB_HAVE_SSE41
         return _mm_ceil_ps(item);
-#elif defined(DLIB_HAVE_SSE2)
+#elif defined(DLIB_HAVE_SSE2) || defined(DLIB_HAVE_NEON)
+      std::cout << "ceil!" << std::endl;
```
Comment (StatusReport) [inline dlib/simd/simd4f.h]: Remove.

File: dlib/simd/simd4f.h [hunk 12]
```diff
@@ -448,7 +571,8 @@ namespace dlib
     {
 #ifdef DLIB_HAVE_SSE41
         return _mm_floor_ps(item);
-#elif defined(DLIB_HAVE_SSE2)
+#elif defined(DLIB_HAVE_SSE2) || defined(DLIB_HAVE_NEON)
+      std::cout << "floor!" << std::endl;
```
Comment (StatusReport) [inline dlib/simd/simd4f.h]: Remove.

**Review by StatusReport — CHANGES_REQUESTED** [review]
Are you sure this is the branch with the latest changes?

File: dlib/simd/simd4f.h (hunk 1 above)
Comment (gershonho) [reply dlib/simd/simd4f.h]: Done.

File: dlib/simd/simd4f.h (hunk 11 above)
Comment (gershonho) [reply dlib/simd/simd4f.h]: Done.

File: dlib/simd/simd4f.h (hunk 12 above)
Comment (gershonho) [reply dlib/simd/simd4f.h]: Done.

File: dlib/simd/simd4f.h (hunk 10 above)
Comment (gershonho) [reply dlib/simd/simd4f.h:491]: Done.

File: dlib/simd/simd4i.h (hunk 9 above)
Comment (gershonho) [reply dlib/simd/simd4i.h]: Done.

File: dlib/simd/simd4i.h (hunk 8 above)
Comment (gershonho) [reply dlib/simd/simd4i.h]: Done.

File: dlib/simd/simd4i.h (hunk 6 above)
Comment (gershonho) [reply dlib/simd/simd4i.h:194]: Correct.

File: dlib/simd/simd4i.h (hunk 5 above)
Comment (gershonho) [reply dlib/simd/simd4i.h]: Done.

File: dlib/simd/simd4i.h (hunk 4 above)
Comment (gershonho) [reply dlib/simd/simd4i.h]: Done.

File: dlib/simd/simd4i.h (hunk 3 above)
Comment (gershonho) [reply dlib/simd/simd4i.h]: Done.

File: dlib/simd/simd4i.h (hunk 2 above)
Comment (gershonho) [reply dlib/simd/simd4i.h]: Done.

**Comment by gershonho** [comment]
Done. This is the latest version (it contains simd4f_bool based on uint32x4_t).

File: dlib/simd/simd4f.h [hunk 13]
```diff
@@ -352,6 +458,12 @@ namespace dlib
     { 
 #ifdef DLIB_HAVE_SSE2
         return _mm_rcp_ps(item); 
+#elif defined(DLIB_HAVE_NEON)
+		float _lhs[4]; item.store(_lhs);
+		return simd4f(1.0f/_lhs[0],
+					  1.0f/_lhs[1],
```
Comment (StatusReport) [inline dlib/simd/simd4f.h]: Indentation.

File: dlib/simd/simd4f.h [hunk 14]
```diff
@@ -366,6 +478,11 @@ namespace dlib
     { 
 #ifdef DLIB_HAVE_SSE2
         return _mm_rsqrt_ps(item); 
+#elif defined(DLIB_HAVE_NEON)
+      float32x4_t x1 = vmaxq_f32(item, vdupq_n_f32(FLT_MIN));
```
Comment (StatusReport) [inline dlib/simd/simd4f.h]: Indentation.

File: dlib/simd/simd4f.h [hunk 15]
```diff
@@ -475,6 +597,12 @@ namespace dlib
         return _mm_blendv_ps(b,a,cmp);
 #elif defined(DLIB_HAVE_SSE2)
         return _mm_or_ps(_mm_and_ps(cmp,a) , _mm_andnot_ps(cmp,b));
+#elif defined(DLIB_HAVE_NEON)
+		return (float32x4_t)vorrq_s32(
```
Comment (StatusReport) [inline dlib/simd/simd4f.h]: Indentation.

File: dlib/simd/simd4i.h (hunk 6 above)
Comment (StatusReport) [reply dlib/simd/simd4i.h:194]: So why do we have a diff here?

File: dlib/simd/simd4f.h [hunk 16]
```diff
@@ -388,6 +505,9 @@ namespace dlib
         simd4f temp = _mm_add_ps(item,_mm_movehl_ps(item,item));
         simd4f temp2 = _mm_shuffle_ps(temp,temp,1);
         return _mm_cvtss_f32(_mm_add_ss(temp,temp2));
+#elif defined(DLIB_HAVE_NEON) //TODO : optz
```
Comment (StatusReport) [inline dlib/simd/simd4f.h]: remove comment

File: dlib/simd/simd4f.h [hunk 17]
```diff
@@ -352,6 +458,12 @@ namespace dlib
     { 
 #ifdef DLIB_HAVE_SSE2
         return _mm_rcp_ps(item); 
+#elif defined(DLIB_HAVE_NEON)
+		float _lhs[4]; item.store(_lhs);
```
Comment (StatusReport) [inline dlib/simd/simd4f.h]: Indentation + newline.

**Review by StatusReport — CHANGES_REQUESTED** [review]
Round.

File: dlib/simd/simd4f.h (hunk 17 above)
Comment (gershonho) [reply dlib/simd/simd4f.h]: Done.

File: dlib/simd/simd4f.h (hunk 13 above)
Comment (gershonho) [reply dlib/simd/simd4f.h]: Done.

File: dlib/simd/simd4f.h (hunk 14 above)
Comment (gershonho) [reply dlib/simd/simd4f.h]: Done.

File: dlib/simd/simd4f.h (hunk 16 above)
Comment (gershonho) [reply dlib/simd/simd4f.h]: Done.

File: dlib/simd/simd4f.h (hunk 15 above)
Comment (gershonho) [reply dlib/simd/simd4f.h]: Done.

File: dlib/simd/simd4i.h (hunk 6 above)
Comment (gershonho) [reply dlib/simd/simd4i.h:194]: The old version has a trailing blank. The new version has this trailing blank dropped (presumably by Xcode).

File: dlib/simd/simd4i.h (hunk 7 above)
Comment (gershonho) [reply dlib/simd/simd4i.h:174]: Same.

**Comment by StatusReport** [comment]
Squash.

**Comment by gershonho** [comment]
Done.

**Review by StatusReport — APPROVED** [review]

**Pull request merged**